src/category_theory/sites/sieves.lean

```diff
@@ -0,0 +1,250 @@
+{ arrows := λ Y, {f | ∃ S ∈ 𝒮, sieve.arrows S f},
+  downward_closed := λ Y Z f, by { rintro ⟨S, hS, hf⟩ g, exact ⟨S, hS, S.downward_closed hf _⟩ } }
+
+/-- The infimum of a collection of sieves: the intersection of them all. -/
+protected def Inf (𝒮 : set (sieve X)) : (sieve X) :=
+{ arrows := λ Y, {f | ∀ S ∈ 𝒮, sieve.arrows S f},
+  downward_closed := λ Y Z f hf g S H, S.downward_closed (hf S H) g }
+
+/-- The union of two sieves is a sieve. -/
+protected def union (S R : sieve X) : sieve X :=
+{ arrows := λ Y f, S.arrows f ∨ R.arrows f,
+  downward_closed := by { rintros Y Z f (h | h) g; simp [h] } }
+
+/-- The intersection of two sieves is a sieve. -/
+protected def inter (S R : sieve X) : sieve X :=
+{ arrows := λ Y f, S.arrows f ∧ R.arrows f,
+  downward_closed := by { rintros Y Z f ⟨h₁, h₂⟩ g, simp [h₁, h₂] } }
+
+/--
+Sieves on an object `X` form a complete lattice.
+We generate this directly rather than using the galois insertion for nicer definitional
+properties.
+-/
+instance : complete_lattice (sieve X) :=
+{ le           := λ S R, ∀ Y (f : Y ⟶ X), S.arrows f → R.arrows f,
+  le_refl      := λ S f q, id,
+  le_trans     := λ S₁ S₂ S₃ S₁₂ S₂₃ Y f h, S₂₃ _ _ (S₁₂ _ _ h),
+  le_antisymm  := λ S R p q, sieve.ext (λ Y f, ⟨p _ _, q _ _⟩),
+  top          := { arrows := λ _, set.univ, downward_closed := λ Y Z f g h, ⟨⟩ },
+  bot          := { arrows := λ _, ∅, downward_closed := λ _ _ _ p _, false.elim p },
+  sup          := sieve.union,
+  inf          := sieve.inter,
+  Sup          := sieve.Sup,
+  Inf          := sieve.Inf,
+  le_Sup       := λ 𝒮 S hS Y f hf, ⟨S, hS, hf⟩,
+  Sup_le       := λ ℰ S hS Y f, by { rintro ⟨R, hR, hf⟩, apply hS R hR _ _ hf },
+  Inf_le       := λ _ _ hS _ _ h, h _ hS,
+  le_Inf       := λ _ _ hS _ _ hf _ hR, hS _ hR _ _ hf,
+  le_sup_left  := λ _ _ _ _, or.inl,
+  le_sup_right := λ _ _ _ _, or.inr,
+  sup_le       := λ _ _ _ a b _ _ hf, hf.elim (a _ _) (b _ _),
+  inf_le_left  := λ _ _ _ _, and.left,
+  inf_le_right := λ _ _ _ _, and.right,
+  le_inf       := λ _ _ _ p q _ _ z, ⟨p _ _ z, q _ _ z⟩,
+  le_top       := λ _ _ _ _, trivial,
+  bot_le       := λ _ _ _, false.elim }
+
+/-- The maximal sieve always exists. -/
+instance sieve_inhabited : inhabited (sieve X) := ⟨⊤⟩
+
+@[simp]
+lemma mem_Inf {Ss : set (sieve X)} {Y} (f : Y ⟶ X) :
+  (Inf Ss).arrows f ↔ ∀ S ∈ Ss, sieve.arrows S f :=
+iff.rfl
+
+@[simp]
+lemma mem_Sup {Ss : set (sieve X)} {Y} (f : Y ⟶ X) :
+  (Sup Ss).arrows f ↔ ∃ S ∈ Ss, sieve.arrows S f :=
+iff.rfl
+
+@[simp]
+lemma mem_inter {R S : sieve X} {Y} (f : Y ⟶ X) :
+  (R ⊓ S).arrows f ↔ R.arrows f ∧ S.arrows f :=
+iff.rfl
+
+@[simp]
+lemma mem_union {R S : sieve X} {Y} (f : Y ⟶ X) :
+  (R ⊔ S).arrows f ↔ R.arrows f ∨ S.arrows f :=
+iff.rfl
+
+@[simp]
+lemma mem_top (f : Y ⟶ X) : (⊤ : sieve X).arrows f := trivial
+
+/-- Take the downward-closure of a set of morphisms to `X`. -/
+inductive generate_sets (𝒢 : set (over X)) : Π (Y : C), set (Y ⟶ X)
```

Comment:
Shouldn't this take something of the same type as `arrow_set` as argument `𝒢` for homogeneity?
It might be good to write an explicit little API for moving back and forth between `set (over X)` and arrow_sets.

Reply:
I think I tried that originally and it made the definition of the galois insertion more awkward - I'll try again though

Reply:
Still no luck - if someone else wants to change the argument type for homogeneity feel free; I think I'm misunderstanding something